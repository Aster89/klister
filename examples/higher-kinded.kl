#lang "prelude.kl"

(import "pair-datatype.kl")


-- Making sure we can write down unsaturated primitive type constructors,
-- primitive datatypes, and user-defined datatypes.

(datatype (OfUnit F)
  (of-unit (F Unit)))

(example
  (the (OfUnit (IO))
       (of-unit (pure-IO unit))))
(example
  (the (OfUnit (-> Unit))
       (of-unit id)))
(example
  (the (OfUnit (Maybe))
       (of-unit (just unit))))
(example
  (the (OfUnit (Pair Unit))
       (of-unit (pair unit unit))))


-- Make sure it also works at kind * -> * -> *.

(datatype (OfUnitUnit F)
  (of-unit-unit
    (F Unit Unit)))

(example
  (the (OfUnitUnit (->))
       (of-unit-unit id)))
(example
  (the (OfUnitUnit (Pair))
       (of-unit-unit (pair unit unit))))


(export
  OfUnit of-unit
  OfUnitUnit of-unit-unit)
