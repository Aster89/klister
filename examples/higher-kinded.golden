(of-unit #<IO action>) : (OfUnit IO)
(of-unit #<closure>) : (OfUnit ((→) Unit))
(of-unit (just (unit))) : (OfUnit Maybe)
(of-unit (pair (unit) (unit))) : (OfUnit (Pair Unit))
(of-unit-unit #<closure>) : (OfUnitUnit (→))
(of-unit-unit #<closure>) : (OfUnitUnit (→))
(of-unit-unit (pair (unit) (unit))) : (OfUnitUnit Pair)
